crypto/fipsmodule/cipher/e_aes.c

```diff
@@ -715,6 +715,63 @@ static int aes_xts_ctrl(EVP_CIPHER_CTX *c, int type, int arg, void *ptr) {
   return 1;
 }
 
+typedef struct {
+  union {
+      double align;
+      AES_KEY ks;
+  } ks;
+  // Indicates if IV has been set.
+  const uint8_t *iv;
+} EVP_AES_WRAP_CTX;
+
+static int aes_wrap_init_key(EVP_CIPHER_CTX *ctx, const uint8_t *key,
+                            const uint8_t *iv, int enc) {
+  EVP_AES_WRAP_CTX *wctx = ctx->cipher_data;
+  if (iv == NULL && key == NULL) {
+    return 1;
+  }
+  if (key != NULL) {
+    if (ctx->encrypt) {
+      AES_set_encrypt_key(key, EVP_CIPHER_CTX_key_length(ctx) * 8,
+                          &wctx->ks.ks);
+    } else {
+      AES_set_decrypt_key(key, EVP_CIPHER_CTX_key_length(ctx) * 8,
+                          &wctx->ks.ks);
+    }
+    if (iv == NULL) {
+      wctx->iv = NULL;
+    }
+  }
+  if (iv != NULL) {
+    memcpy(ctx->iv, iv, ctx->cipher->iv_len);
+    wctx->iv = ctx->iv;
+  }
+  return 1;
+}
+
+static int aes_wrap_cipher(EVP_CIPHER_CTX *ctx, uint8_t *out, const uint8_t *in,
+                           size_t inlen) {
+  EVP_AES_WRAP_CTX *wctx = ctx->cipher_data;
+  // There is no final operation, so we always return zero length here.
+  if (in == NULL) {
+    return 0;
+  }
+
+  // Internal calls to |AES_wrap/unwrap_key| within the fipsmodule should be
+  // wrapped with state lock functions to avoid updating the service indicator.
+  // When consuming via |EVP_CIPHER|, |EVP_CipherFinal(_ex)| should be the
+  // function that indicates approval.
+  int ret;
+  FIPS_service_indicator_lock_state();
+  if (ctx->encrypt) {
+    ret = AES_wrap_key(&wctx->ks.ks, wctx->iv, out, in, inlen);
+  } else {
+    ret = AES_unwrap_key(&wctx->ks.ks, wctx->iv, out, in, inlen);
+  }
+  FIPS_service_indicator_unlock_state();
+  return ret;
```

Comment:
Why do some EVP_CIPHER.cipher return success 0/1 (XTS and CBC) and others return a length (GCM and now key wrap)?